### ISSUE TITLE
Download URLs Missing Subdirectories

### ISSUE THREAD

**Issue by thedward** [erickok/transdroid#518]
Transdroid: 2.5.16
Transmission: 2.94

When using the ftp(s) download feature, the constructed URLs don't include any subdirectories.

For example: if my root download directory is `/home/user/files/` and my base ftp(s) URL is `ftps://user@server/private/files/` and I have a torrent saved in `/home/user/files/books` that contains a file `book.epub` when I click the download button for `book.epub` it tries to download from the URL `ftps://user@server/private/files/book.epub` instead of the expected `ftps://user@server/private/files/books/book.epub`.

If I move the torrent to the root download directory, it works as expected.

Is this a bug, or am I doing something wrong?

**Comment by erickok** [comment]
Which torrent client are you using?

Op do 12 dec. 2019 21:34 schreef Thedward Blevins <notifications@github.com

**Comment by thedward** [comment]
@erickok Transmission 2.94

**Comment by erickok** [comment]
Ok that makes sense. Given a base url `ftps://user@server/private/files/` it will just append that torrent relative path, so `book.epub` since that is the torrent's location (`/`) appended with the torrent's file's relative path (`book.epub`). Unless the torrent itseld declares the `books` directory, I am not sure how/where the `books` comes from?

**Comment by thedward** [comment]
* Transmission's download directory is set to `/home/user/files`
* This particular torrent was downloaded to `books` (`/home/user/files/books`)
* The URL `ftps://user@server/private/files` serves up the directory `/home/user/files`
* Since the torrent's download directory relative to transmission's root download directory is `books`, I would expect the string appended to the ftp URL to be `/books/book.epub` instead of `/book. epub`
* Basically it only works correctly for torrents saved directly in transmission's root download directory.

**Comment by erickok** [comment]
Yes I am sorry but since Transmission doesn't report (or even know?) that the files are in a certain subdirectory, there is no way afaik to fix this.

**Comment by thedward** [comment]
That is incorrect. Transmission will report the download directory via the `downloadDir` field for each torrent in response to an appropriate `torrent-get` RPC request.

**Comment by erickok** [comment]
I believe I use that already. I will give it another look.

**Comment by thedward** [comment]
In `org/transdroid/core/gui/DetailsFragment.java` the download URL is constructed using the `relativePath` attribute (via `getRelativePath`) of the `TorrentFile`.  For Transmission at least, `TorrentFile.relativePath` is relative to `Torrent.locationDir`, so that means the constructed URL will only be correct if  `ServerSettings.ftpUrl` is set to a URL that is mapped to `Torrent.locationDir` on the server. Generally, a given URL will only map to one directory, so if you have different torrents downloaded to different directories, this functionality will currently only work for torrents downloaded to a single particular directory. 

Instead, I would expect it to work something like this:

- `ServerSettings.ftpUrl` is assumed to map to the daemon's default download path.
- The download file option is only be available when  `Torrent.locationDir` is a subpath of (or the same as) the daemon's default download path.
- The download URL is constructed by removing (if present) the string representing the daemon's default download path from the beginning of  `TorrentFile.fullPath` and appending the resulting relative URL to `ServerSettings.ftpUrl`

**Comment by erickok** [comment]
Yes. Transmission report the torrent location in the `locationDir`, which is an absolute path (f.e. `/path/to/dowloads/books/`), and the files are reported with their dir/filename relative to this (f.e. `My.Torrent/Some.File.txt`). So to determine the complete relative path (f.e. `/books/My.Torrent/Some.File.txt`, which is required for a proper ftp download link) we need to know to what location the ftp site points to. I have always assumed this is the torren'ts base location, but obviously that is not correct.

**Comment by erickok** [comment]
Alright, for the next release I allowed for Transmission servers to set the base download dir (under Optional Settings -> Download directory). If you set this to your absolute download dir (f.e. `/home/user/files` without trailing slash) then that woulds be used to make a best guess at the relative url to use for downloads.
